### DISCUSSION TITLE
Does famfs provide POSIX-like interfaces for operating files under the famfs file system, such as open, close, write, read, lseek, and other functions?

### DISCUSSION THREAD

**Discussion by lordiscat** [cxl-micron-reskit/famfs#52] [Q&A] [answered]
Does famfs provide POSIX-like interfaces for operating files under the famfs file system, such as open, close, write, read, lseek, and other functions?

**Comment by jagalactic** [comment, marked as answer]
Yes, with some limitations:

The most interesting use case for famfs is ```mmap()```. If you memory map a famfs file, you are directly accessing the memory. Unlike conventional mmap, famfs does not demand-page data from a backing device (e.g. SSD) into the page cache. If you access an address from a memory mapped famfs file, the memory is accessed directly; cache lines are loaded into the processor cache directly from the memory.

Back to the question, If a file already exists, you can use any of the posix calls, with the following limitations:

- Writing works up to, but not past EOF, because famfs does not allocate-on-write
- Truncate is not allowed - but famfs can't currently prevent it, so a file whose size has been modified will be marked invalid (and stop working). Recovery from this is umount/remount.

Creating a file is special, because famfs never does allocate-on-write. So you need to use the famfs CLI or API:

- ```famfs creat -s <size> <path>```
- ```famfs cp <srcfile> <destfile>```    #srcfile can be anywhere; destfile is in a mounted famfs file system
- You can also create files via the famfs library that is called by the cli

One way to think about it is that creating a famfs file is analogous to allocating memory that can be referenced via the file. 

Also, if you are using a shared-memory dax device, only the host that created the filesystem can create files. This is because both memory allocation and metadata log append must be serialized. 

Many of the current limitations can be relaxed or eliminated if use cases justify adding complexity to famfs.

Hope this helps. Let me know if you have more questions.

John
